### PR TITLE
New feature added -> combat challenges

### PR DESCRIPTION
**Pull request by nikolalsvk**

### CODE REVIEW

File: app/controllers/char_combats_controller.rb [hunk 1]
```diff
@@ -38,7 +51,19 @@ def find_second_char
...
+  def find_char_type
+    @char_type = current_user.char_types.find_by_id(params["char_type"])
+  end
+
   def find_char_combat
     @char_combat = CharCombat.find_by_id(params[:id])
   end
+
+  def combat_challenges
+    CharCombat.where("challenge = ?", true)
```
Comment (pyromaniackeca) [inline app/controllers/char_combats_controller.rb]: You can write this as: `where(:challenge => true)`

File: app/models/char_combat.rb [hunk 2]
```diff
@@ -3,14 +3,20 @@ class CharCombat < ActiveRecord::Base
   belongs_to :first_combatant, :class_name => CharType.to_s
   belongs_to :second_combatant, :class_name => CharType.to_s
 
-  validates :first_combatant, :presence => true
-  validates :second_combatant, :presence => true
+  validates :first_combatant, :presence => true, :if => lambda { second_combatant == nil }
```
Comment (pyromaniackeca) [inline app/models/char_combat.rb]: You can write a function like `def smth . . . end` and pass it like this: `:if => :smth`.

File: app/models/char_combat.rb [hunk 3]
```diff
@@ -3,14 +3,20 @@ class CharCombat < ActiveRecord::Base
...
+  validates :second_combatant, :presence => true, :if => lambda { first_combatant == nil }
 
-  before_create :combat
+  before_save :open_challenge, :combat
+
+  def open_challenge
+    self.challenge = second == nil
+    nil
```
Comment (pyromaniackeca) [inline app/models/char_combat.rb:13]: Why are these functions returning nil?

File: app/views/char_types/index.html.erb [hunk 4]
```diff
@@ -43,7 +43,8 @@
             </tr>
           </table>
         <% else %>
-          <p>This character type has no attributes! <%= link_to 'Add some', new_char_type_char_attribute_path(type) %>.</p>
+          <p>This character type has no attributes! <%= link_to 'Add some',
```
Comment (pyromaniackeca) [inline app/views/char_types/index.html.erb]: I think this is better as one line, either that or put the whole link in the next line.

File: db/migrate/20150804140800_create_char_combats.rb [hunk 5]
```diff
@@ -4,6 +4,7 @@ def change
       t.integer :first_combatant_id
       t.integer :second_combatant_id
       t.integer :winner
+      t.boolean	:challenge
```
Comment (pyromaniackeca) [inline db/migrate/20150804140800_create_char_combats.rb]: Fix indentation.

File: features/combat_challenge.feature [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+    And there is open challenge
+
+  Scenario: I view open challenge
+    When I go to my character page
+    Then I should see open challenge
+
+  Scenario: I join open challenge
+    When I go to my character page
+    And I click on "Join" link
+    Then I should see who won the challenge
```
Comment (pyromaniackeca) [inline features/combat_challenge.feature]: Add new line.

File: app/models/char_combat.rb (hunk 3 above)
Comment (nikolalsvk) [inline app/models/char_combat.rb:13]: @pyromaniackeca Explanation -> http://apidock.com/rails/ActiveRecord/RecordNotSaved

File: app/controllers/char_combats_controller.rb [hunk 7]
```diff
@@ -29,6 +30,18 @@ def destroy
     redirect_to char_combats_path
   end
 
+  def edit
```
Comment (nikolalsvk) [inline app/controllers/char_combats_controller.rb]: Change this to update.

File: app/models/char_type.rb [hunk 8]
```diff
@@ -29,8 +29,13 @@ def attack_points
...
   def get_combat_history
-    CharCombat.where("first_combatant_id = ? OR second_combatant_id = ?",
-                     self, self)
+    CharCombat.where("(first_combatant_id = ? OR second_combatant_id = ?) and challenge = ?",
+                     self, self, false)
+  end
+
+  def get_challenges
+    CharCombat.where("(first_combatant_id != ? OR second_combatant_id != ?) and challenge = ?",
+                     self, self, true)
```
Comment (nikolalsvk) [inline app/models/char_type.rb]: Make this more DRY.

File: app/views/char_combats/_first.html.erb [hunk 9]
```diff
@@ -0,0 +1,5 @@
+<% if char_combat.first_combatant == nil %>
+  <f></f>
+<% else %>
+  <%= char_combat.first_combatant.title %>
+<% end %>
```
Comment (nikolalsvk) [inline app/views/char_combats/_first.html.erb]: Add a new line here.

File: features/combat_challenge.feature [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+
```
Comment (nikolalsvk) [inline features/combat_challenge.feature]: Add a new line.

File: app/controllers/char_combats_controller.rb [hunk 11]
```diff
@@ -29,6 +30,18 @@ def destroy
     redirect_to char_combats_path
   end
 
+  def update
+    find_char_type
```
Comment (pyromaniackeca) [inline app/controllers/char_combats_controller.rb]: You can put this as a before_action.

File: app/models/char_type.rb [hunk 12]
```diff
@@ -28,12 +28,16 @@ def attack_points
...
   end
 
-  def destroy_char_combat_history
-    get_combat_history.delete_all
+  def combats
+    CharCombat.where("first_combatant_id != ? OR second_combatant_id != ?", 
+          self, self)
+  end
+
+  def challenges(bool)
```
Comment (pyromaniackeca) [inline app/models/char_type.rb]: Actually I meant for this to be more like:

```
def challenges
  comabats.where(:challenge => true)
end

def combat_history
  combats.where(:challenge => false)
end
```

File: features/combat_challenge.feature (hunk 10 above)
Comment (pyromaniackeca) [inline features/combat_challenge.feature]: Still no new line.

**Comment by pyromaniackeca** [comment]
If it passes on Semaphore it's ok to merge.

**Pull request merged**